### PR TITLE
Allow for more flexible layouts in `layout.jelly`

### PR DESCRIPTION
**Pull request by janfaracik**
**Background**

Currently in Jenkins you're limited to three layout types - `two-column`, `one-column` and `full-screen`. This PR reworks `layout.jelly` to allow for more layout types, for example a three column layout or a layout with the side panel on the right.

This gives developers more flexibility in how they approach the design of their pages and also simplifies how `layout.jelly` and it's related stylings work.

### Testing done

* Pages appear as before, with some minor padding 
* Full screen pages appear as they did before
* Mobile view/breakpoints behave in the same way
* Design Library's title is currently broken - ~I'll raise a PR to fix that~ PR has been raised https://github.com/jenkinsci/design-library-plugin/pull/315

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jenkinsci/sig-ux 

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly [hunk 1]
```diff
@@ -27,8 +27,7 @@ THE SOFTWARE.
 -->
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-<l:layout permissions="${app.MANAGE_AND_SYSTEM_READ}" title="${%System}" type="one-column">
```
Comment (janfaracik) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly:30]: One issue with the current approach is that there can be a mismatch between the page type and it's contents - on the `Configure` page it includes a sidebar however it isn't displayed due to the type being set to `one-column`. The inverse could be true as well where the page type is `two-column` but a side panel hasn't been provided.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 2]
```diff
@@ -58,7 +58,7 @@ THE SOFTWARE.
       (The permissions will be checked against the "it" object.)
     </st:attribute>
     <st:attribute name="type" use="optional">
-      Available values: two-column (by default), one-column (full-width size) or full-screen (since 2.53).
```
Comment (janfaracik) [inline core/src/main/resources/lib/layout/layout.jelly:61]: The `type` attribute now only works if it's set to `full-screen`.

File: war/src/main/scss/components/_section.scss [hunk 3]
```diff
@@ -3,7 +3,6 @@
 .jenkins-section {
   border-top: 2px solid var(--panel-border-color);
   padding: var(--section-padding) 0 0 0;
-  max-width: 1800px;
```
Comment (janfaracik) [inline war/src/main/scss/components/_section.scss:6]: Simplified the styling, let's just fill the available width and let the parent container control how wide that can be.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 4]
```diff
@@ -191,14 +191,8 @@ THE SOFTWARE.
     </j:if>
 
     <div id="page-body" class="app-page-body app-page-body--${layoutType} clear">
-      <j:if test="${layoutType=='two-column'}">
-        <j:set var="mode" value="side-panel" />
-        <d:invokeBody />
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/layout.jelly:196]: Breaks all context menus.

**Review by daniel-beck — CHANGES_REQUESTED** [review]

**Comment by daniel-beck** [comment]
This PR looks a lot like

> For new APIs and extension points, there is a link to at least one consumer.

would apply (on a UI level), but I don't see consumers for this change?

File: core/src/main/resources/lib/layout/layout.jelly (hunk 4 above)
Comment (janfaracik) [reply core/src/main/resources/lib/layout/layout.jelly:196]: Thanks! Fixed it and added a comment to `side-panel.jelly`

**Comment by janfaracik** [comment]
> This PR looks a lot like
> 
> > For new APIs and extension points, there is a link to at least one consumer.
> 
> would apply (on a UI level), but I don't see consumers for this change?

I've added some guidance on using the new layout system in the Design Library - https://github.com/jenkinsci/design-library-plugin/pull/315 - intending to update the UI of the plugin itself to use a multi column layout, however I haven't had the time recently.

**Comment by timja** [comment]
> > This PR looks a lot like
> > > For new APIs and extension points, there is a link to at least one consumer.
> > 
> > 
> > would apply (on a UI level), but I don't see consumers for this change?
> 
> I've added some guidance on using the new layout system in the Design Library - [jenkinsci/design-library-plugin#315](https://github.com/jenkinsci/design-library-plugin/pull/315) - intending to update the UI of the plugin itself to use a multi column layout, however I haven't had the time recently.

possible to update the UI to use it? should really have a consumer before shipping something like this

**Comment by janfaracik** [comment]
> possible to update the UI to use it? should really have a consumer before shipping something like this

Will do 👍

**Comment by janfaracik** [comment]
> possible to update the UI to use it? should really have a consumer before shipping something like this

Updated the Design Library with an additional side-panel which gives an overview of the page https://github.com/jenkinsci/design-library-plugin/pull/315

image

**Comment by daniel-beck** [comment]
> > possible to update the UI to use it? should really have a consumer before shipping something like this
> 
> Updated the Design Library with an additional side-panel which gives an overview of the page [jenkinsci/design-library-plugin#315](https://github.com/jenkinsci/design-library-plugin/pull/315)

I don't think the UI demo plugin qualifies as a consumer that would validate that this change does anything useful.

**Comment by timja** [comment]
Paused for now

**Pull request closed without merging**